### PR TITLE
Enable sorting when batching is enabled

### PR DESCRIPTION
**Pull request by PaulSchweizer**
This is a followup of our discussion from July 18th about batching and sorting. 
I took a deeper dive into the batching feature and have a better understanding of it now.
Basically the core issue was, that the batching resolver did not take care of running the initial query. Instead, the initial query had to be run "manually" in a custom "resolve_" field as indicated in the tests: https://github.com/graphql-python/graphene-sqlalchemy/blob/dfee3e9417cdb8a6ec67b5cd79ee203ce4f72ed7/graphene_sqlalchemy/tests/test_batching.py#L67
That of course bypassed the "get_query" function in the resolver and thus also made sorting "impossible".
Let me know what you think.

### CODE REVIEW

File: graphene_sqlalchemy/batching.py
Comment (PaulSchweizer) [inline graphene_sqlalchemy/batching.py:13]: We store the loaders per relationship so each item accesses the same cached result ensuring that the selectin is only triggered once

File: graphene_sqlalchemy/batching.py
Comment (PaulSchweizer) [inline graphene_sqlalchemy/batching.py:105]: I'm not 100% sure about this. Do any of you see issues with this?

File: graphene_sqlalchemy/fields.py
Comment (PaulSchweizer) [inline graphene_sqlalchemy/fields.py:132]: I'm not sure what our exact plan in regards to unifying these fields was. I just made this sortable, we can discuss this again

File: graphene_sqlalchemy/tests/models.py
Comment (PaulSchweizer) [inline graphene_sqlalchemy/tests/models.py:127]: Had to add some deeper nesting in the test data for more robust tests

File: graphene_sqlalchemy/tests/test_batching.py
Comment (PaulSchweizer) [inline graphene_sqlalchemy/tests/test_batching.py:101]: As mentioned, the original test schema defines custom resolve_ methods which I find odd, but maybe people use it like that. We use it like this.

File: graphene_sqlalchemy/tests/test_batching.py
Comment (PaulSchweizer) [inline graphene_sqlalchemy/tests/test_batching.py:123]: unfortunately there were some flake8 errors I had to fix, so ignore these indentation changes

File: graphene_sqlalchemy/batching.py
Comment (erikwrede) [reply graphene_sqlalchemy/batching.py:105]: Why do we need the get_event_loop() call in the else part? Isn't that initially covered by `aiodataloader` already, and shouldn't the cache also get a full reset if the event loop changes (implying thread restart/change)?

`aiodataloader` recommends to create a new loader for each request, this would keep the loaders between different queries, right?

File: graphene_sqlalchemy/fields.py
Comment (erikwrede) [reply graphene_sqlalchemy/fields.py:132]: Looks good.
If I'm correct, this would remove the need for the separate unsorted field as sorting for all the possible fields could be disabled using the `Meta` options. Maybe there is a need to add a QoL `sort=None` shortcut for the  options.

**Comment by erikwrede** [comment]
https://github.com/graphql-python/graphene-sqlalchemy/blob/ac57fd4ba78a786f36aa0c948596a66766c33ec7/graphene_sqlalchemy/enums.py#L147

Changing this to `field_name` would fix the errors with sqa-13.
It seems like `column_prop` of `Reporter` has the name set to something similar to `%(4447102096 anon)s` at the time of sort enum generation. This causes `def sort_enum_for_object_type` to generate a sort enum with field names not allowed by GQL (`%(4447102096 anon)s_ASC  & %(4447102096 anon)s_DESC`), causing the test failure. Couldn't figure out why `get_schema` does not reproduce the same behavior although the same models were used.

File: graphene_sqlalchemy/batching.py
Comment (PaulSchweizer) [reply graphene_sqlalchemy/batching.py]: Makse sense, I changed it

File: graphene_sqlalchemy/fields.py
Comment (PaulSchweizer) [reply graphene_sqlalchemy/fields.py:137]: I deprecated the UnsortedSQLAlchemyConnectionField as you can just set `sort=None` on the meta class already to disable sorting: https://github.com/graphql-python/graphene-sqlalchemy/blob/6417061998823fb3531def8fee394bd81c6f4c24/graphene_sqlalchemy/tests/test_sort_enums.py#L263

**Comment by PaulSchweizer** [comment]
I still need to investigate those failing tests

**Comment by PaulSchweizer** [comment]
The tests pass now, the remaining issue was that sqla1.2 is creating slighlty different select statments so my asserts where not working there. This is ready from my side now, please give it another look.

**Review by erikwrede — CHANGES_REQUESTED** [review]
Tested this locally. Working nicely!
I suggested a few minor adjustments. Please check my comments.

File: graphene_sqlalchemy/batching.py
Comment (PaulSchweizer) [reply graphene_sqlalchemy/batching.py:92]: The cache is necessary because during a request, each `resolve` needs to access the same loader so it can draw on the same cache. Without keeping a reference to the loader between resolves, we would create a new loader for each `resolve` and thus not be able to draw on the loader's cache.

Note though, that the loader is recreated for each request, so it is not really "cached" between requests, the loader object just lingers around in this dict until the next request. I'm not sure how I could clear this cache after the request has been resolved if that is what you where thinking about. You'd have to give me a pointer where and how that could be implemented.

Comment (erikwrede) [reply graphene_sqlalchemy/batching.py:92]: I didn't see that the loader is recreated for each request. I think we can leave it as it is then. The cache not being cleared shouldn't cause any additional trouble.

File: graphene_sqlalchemy/fields.py
Comment (PaulSchweizer) [reply graphene_sqlalchemy/fields.py:126]: I added your suggestion and a test for it.

File: graphene_sqlalchemy/enums.py
Comment (sabard) [reply graphene_sqlalchemy/enums.py:147]: This change looks reasonable to me. The only place I can see there being an issue is if a user defined a custom `ORMField` attribute on the ObjectType or any parent class (see [types.py#L128](https://github.com/graphql-python/graphene-sqlalchemy/blob/515060b6ea37f2fa32855bfd8a4c8c49bf5bc2e8/graphene_sqlalchemy/types.py#L128)) with a different name from the field it references. I can write a test case to see what happens in this case or we can just include a note in the release.

File: graphene_sqlalchemy/tests/models.py
Comment (sabard) [reply graphene_sqlalchemy/tests/models.py:127]: I added a `Tags` class in #357. Will see if there's a way to consolidate over there after this is merged.

**Review by sabard — COMMENTED** [review]
Aside from Erik's comments, all LGTM! Let me know if you'd like me to write a test case for custom `ORMField` attributes.

**Comment by PaulSchweizer** [comment]
It would be great @sabard, if you could write the test for the potential issue you have in mind. If it fails I will look into it and work on a solution.

File: graphene_sqlalchemy/tests/test_batching.py
Comment (sabard) [inline graphene_sqlalchemy/tests/test_batching.py:569]: This works as expected as far as I can tell. You can sort with `FIRSTNAME_*` or `FIRST_NAME_*` which is still automatically created.

File: graphene_sqlalchemy/tests/test_batching.py
Comment (sabard) [inline graphene_sqlalchemy/tests/test_batching.py:629]: Do you all have any idea why this results in 2 select statements? Seems to be a relay thing. Full SQL if helpful:

```SQL
SELECT count(*) AS count_1 
FROM (SELECT (SELECT CAST(count(reporters.id) AS INTEGER) AS count_2 
FROM reporters) AS anon_2, reporters.id AS reporters_id, reporters.first_name AS reporters_first_name, reporters.last_name AS reporters_last_name, reporters.email AS reporters_email, reporters.favorite_pet_kind AS reporters_favorite_pet_kind 
FROM reporters ORDER BY reporters.first_name DESC) AS anon_1

SELECT (SELECT CAST(count(reporters.id) AS INTEGER) AS count_1 
FROM reporters) AS anon_1, reporters.id AS reporters_id, reporters.first_name AS reporters_first_name, reporters.last_name AS reporters_last_name, reporters.email AS reporters_email, reporters.favorite_pet_kind AS reporters_favorite_pet_kind 
FROM reporters ORDER BY reporters.first_name DESC
LIMIT ? OFFSET ?

```

**Review by sabard — COMMENTED** [review]
Looks like all works well for custom `ORMField` names.

Off topic, but is sorting supported on non-Relay-style batching? I wasn't able to get it to work; maybe deserves its own test if supported?

File: graphene_sqlalchemy/tests/test_batching.py
Comment (sabard) [reply graphene_sqlalchemy/tests/test_batching.py:629]: Explained offline—first select statement is to determine count.

File: graphene_sqlalchemy/batching.py
Comment (flipbit03) [reply graphene_sqlalchemy/batching.py]: September update: There is also preliminary work in Graphene to vendor the `DataLoader` from `aiodataloader` into Graphene, and also change it's `__init__` behavior to not ask for an even loop so early on init. Just warning because that might touch the same points we are touching here.

**Review by sabard — APPROVED** [review]

**Review by erikwrede — APPROVED** [review]

**Comment by erikwrede** [comment]
@sabard are we ready to merge?

**Pull request merged**
